lib/active_record_shards/shard_selection.rb

```diff
@@ -8,14 +8,65 @@ def initialize
       @on_slave = false
     end
 
-    def shard(klass = nil)
-      if (@shard || self.class.default_shard) && (klass.nil? || klass.is_sharded?)
-        if @shard == NO_SHARD
+    if ActiveRecord::VERSION::MAJOR < 5
+
+      def shard(klass = nil)
+        if (@shard || self.class.default_shard) && (klass.nil? || klass.is_sharded?)
+          if @shard == NO_SHARD
+            nil
+          else
+            @shard || self.class.default_shard
+          end
+        end
+      end
+
+      def shard_name(klass = nil, try_slave = true)
+        the_shard = shard(klass)
+
+        @shard_names ||= {}
+        @shard_names[ActiveRecordShards.rails_env] ||= {}
+        @shard_names[ActiveRecordShards.rails_env][the_shard] ||= {}
+        @shard_names[ActiveRecordShards.rails_env][the_shard][try_slave] ||= {}
+        @shard_names[ActiveRecordShards.rails_env][the_shard][try_slave][@on_slave] ||= begin
+          s = ActiveRecordShards.rails_env.dup
+          s << "_shard_#{the_shard}" if the_shard
+          s << "_slave"              if @on_slave && try_slave
+          s
+        end
+      end
+
+    else
+
+      def shard
+        if @shard.nil? || @shard == NO_SHARD
           nil
         else
           @shard || self.class.default_shard
         end
       end
+
+      PRIMARY = "primary".freeze
+      def resolve_connection_name(sharded:, configurations:)
+        resolved_shard = sharded ? shard : nil
+
+        if !resolved_shard && !@on_slave
+          return PRIMARY
+        end
+
+        @shard_names ||= {}
+        @shard_names[ActiveRecordShards.rails_env] ||= {}
+        @shard_names[ActiveRecordShards.rails_env][resolved_shard] ||= {}
+        @shard_names[ActiveRecordShards.rails_env][resolved_shard][@on_slave] ||= begin
+          s = ActiveRecordShards.rails_env.dup
+          s << "_shard_#{resolved_shard}" if resolved_shard
+
+          if @on_slave && configurations["#{s}_slave"] # fall back to master connection
+            s << "_slave"
+          end
+          s
+        end
```

Comment:
hmmm, these code paths could and maybe should be resolved into one.  I think I got trigger happy thinking we'd just get away with a clean rewrite for AR5... the only AR5 specific line is L53, `return PRIMARY` iirc